java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java

```diff
@@ -0,0 +1,1464 @@
+      if (initialDataBufferSize > 0) {
+        referenceBufferSize = initialDataBufferSize;
+      } else {
+        referenceBufferSize = lastValueAllocationSizeInBytes;
+      }
+      if (dataBuffers.isEmpty()) {
+        // the first data buffer needs to be added
+        ArrowBuf newDataBuf = allocator.buffer(referenceBufferSize);
+        // set length
+        valueBuffer.setInt(writePosition, length);
+        writePosition += LENGTH_WIDTH;
+        // set prefix
+        valueBuffer.setBytes(writePosition, value, start, PREFIX_WIDTH);
+        writePosition += PREFIX_WIDTH;
+        // set buf id
+        valueBuffer.setInt(writePosition, /*first buffer*/0);
+        writePosition += BUF_INDEX_WIDTH;
+        // set offset
+        valueBuffer.setInt(writePosition, 0);
+        newDataBuf.setBytes(0, value, 0, length);
+        newDataBuf.writerIndex(length);
+        dataBuffers.add(newDataBuf);
+      } else {
+        // insert to the last buffer in the data buffers or allocate new if the last buffer isn't enough
+        // set lengths
+        valueBuffer.setInt(writePosition, length);
+        writePosition += LENGTH_WIDTH;
+        // set prefix
+        valueBuffer.setBytes(writePosition, value, start, PREFIX_WIDTH);
+        writePosition += PREFIX_WIDTH;
+        // set buf id
+        int currentBufId = dataBuffers.size() - 1;
+        ArrowBuf currentBuf = dataBuffers.get(currentBufId);
+        if (currentBuf.capacity() - currentBuf.writerIndex() >= length) {
+          // current buffer is enough
+          // set buf indexes
+          valueBuffer.setInt(writePosition, currentBufId);
+          writePosition += BUF_INDEX_WIDTH;
+          // set offset
+          valueBuffer.setInt(writePosition, (int) currentBuf.writerIndex());
+          currentBuf.setBytes(currentBuf.writerIndex(), value, start, length);
+          currentBuf.writerIndex(currentBuf.writerIndex() + length);
+          dataBuffers.set(currentBufId, currentBuf);
+        } else {
+          // current buffer is not enough
+          // allocate new buffer
+          ArrowBuf newBuf = allocator.buffer(referenceBufferSize);
+          // set buf index
+          valueBuffer.setInt(writePosition, dataBuffers.size());
+          writePosition += BUF_INDEX_WIDTH;
+          // set offset
+          valueBuffer.setInt(writePosition, 0);
+          newBuf.setBytes(0, value, start, length);
+          newBuf.writerIndex(newBuf.writerIndex() + length);
+          dataBuffers.add(newBuf);
+        }
+      }
+    }
+  }
+
+  protected final void setBytes(int index, byte[] value, int start, int length) {
+    createViewBuffer(allocator, index, value, start, length, viewBuffer, dataBuffers);
+  }
+
+  /**
+   * Get the total length of the elements up to the given index.
+   * @param index The index of the element in the vector.
+   * @return The total length up to the element at the given index.
+   */
+  public final int getTotalLengthUptoIndex(int index) {
```

Comment:
This is a useful function, but I have doubts about the naming.

Reply:
```suggestion
  public final int getTotalValueLengthUpToIndex(int index) {
```